File: src/leetCode/java/Solution981.java
package leetCode.java;

import java.util.*;

//Binary Search TimeMap(), set(): O(1) O(n), get(): O(logn) O(n)
class TimeMap {//Solution981
    Map<String, List<Object[]>> store;//<key, <[時間, value], ...>>

    public TimeMap() {
        store = new HashMap<>();
    }

    public void set(String key, String value, int timestamp) {
        if(!store.containsKey(key))
            store.put(key, new ArrayList<>());

        store.get(key).add(new Object[]{timestamp, value});
    }

    public String get(String key, int timestamp) {
        if(!store.containsKey(key)) return "";

        List<Object[]> list = store.get(key);
        int l = 0;
        int r = list.size() - 1;
        String res = "";
        while(l <= r){
            int mid = (l + r) / 2;
            int time = (int) list.get(mid)[0];
            if(time == timestamp){
                return (String) list.get(mid)[1];
            }else if(time < timestamp){
                res = (String) list.get(l)[1];//記錄最近的左指標紀錄
                l = mid + 1;
            }else{
                r = mid - 1;
            }
        }
        return res;
    }
}

File: src/leetCode/c++/Solution981.cpp
//Binary Search TimeMap(), set(): O(1) O(n), get(): O(logn) O(n)
class TimeMap {//Solution981
public:
    unordered_map<string, map<int, string>> memo; 

    TimeMap() {        
    }
    
    void set(string key, string value, int timestamp) {
        //all the timestamps are strictly increasing
        memo[key].insert({timestamp, value});
    }
    
    string get(string key, int timestamp) {
        auto it = memo[key].upper_bound(timestamp);//auto=map<int, string>::iterator
        return it == memo[key].begin() ? "" : prev(it)->second;//往前找一個
    }
};
/*
map.upper_bound() 找出大於這個數值的最小值
map.lower_bound() 找出大於或等或等於這個數值的最小值

TreeMap.floorKey() 找出大於或等於這個數值的最小值
TreeMap.ceilingKey 找出小於或等於這個數值的最大值
*/